src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs

```diff
@@ -347,14 +357,125 @@ public async Task CallActivityAsync(DispatchMiddlewareContext dispatchContext, F
                     ResponseEvent = new TaskFailedEvent(
                         eventId: -1,
                         taskScheduledId: scheduledEvent.EventId,
-                        reason: $"Function {functionName} failed with an unhandled exception.",
-                        exceptionDetails),
+                        reason: $"Function '{functionName}' failed with an unhandled exception.",
+                        details: null,
+                        GetFailureDetails(result.Exception)),
                 };
             }
 
             // Send the result of the activity function to the DTFx dispatch pipeline.
             // This allows us to bypass the default, in-process execution and process the given results immediately.
             dispatchContext.SetProperty(activityResult);
         }
+
+#if NET6_0_OR_GREATER
+        private static FailureDetails GetFailureDetails(Exception e)
+        {
+            if (e.InnerException != null && e.InnerException.Message.StartsWith("Result:"))
+            {
+                Exception rpcException = e.InnerException;
+                if (TryGetRpcExceptionFields(rpcException.Message, out string? exception, out string? stackTrace))
+                {
+                    if (TrySplitExceptionTypeFromMessage(exception, out string? exceptionType, out string? exceptionMessage))
+                    {
+                        return new FailureDetails(exceptionType, exceptionMessage, stackTrace, innerFailure: null, isNonRetriable: false);
+                    }
+
+                    return new FailureDetails("(unknown)", exception, stackTrace, innerFailure: null, isNonRetriable: false);
+                }
+                else
+                {
+                    // Don't recognize this message format - return the failure as-is
+                    return new FailureDetails(rpcException);
+                }
+            }
+
+            // Don't recognize this exception - return it as-is
+            return new FailureDetails(e);
+        }
+
+        private static bool TryGetRpcExceptionFields(
+            string rpcExceptionMessage,
+            [NotNullWhen(true)] out string? exception,
+            out string? stackTrace)
+        {
+            exception = null;
+            stackTrace = null;
+
+            // The message string is formatted as follows:
+            // Result: {result}\nException: {message}\nStack: {stack}
+
+            const string ExceptionDelimeter = "\nException: ";
+            const string StackDelimeter = "\nStack: ";
```

Comment:
What happens if a Java user throws an Exception whose message contains the substring `\nException`. Could that throw off our parsing here?

Reply:
For that specific case, no, because we only look for the first instance of `\nException`, which is always going to be the one we want. However, this logic will be thrown off if a Java (or a .NET Isolated) exception message were to contain `\nStack`. I might be able to fix that by using `LastIndexOf(StackDelimeter)` instead of `IndexOf(StackDelimeter)`, in which case we'll only look for the _last_ instance and bypass any that might be embedded in the exception message... Let me post an update for that.

Reply:
Sounds good, I'll await that update, do ping me :)